osquery/tables/system/darwin/certificates.mm

```diff
@@ -66,17 +51,94 @@ void genCertificate(const SecCertificateRef& SecCert, QueryData& results) {
       (cert->skid) ? genKIDProperty(cert->skid->data, cert->skid->length) : "";
 
   results.push_back(r);
-  X509_free(cert);
+}
+
+void genKeychainCertificate(const SecCertificateRef& SecCert,
+                            QueryData& results) {
+  auto der_encoded_data = SecCertificateCopyData(SecCert);
+  if (der_encoded_data == nullptr) {
+    return;
+  }
+
+  const unsigned char* der_bytes = CFDataGetBytePtr(der_encoded_data);
+  auto length = CFDataGetLength(der_encoded_data);
+  auto cert = d2i_X509(nullptr, &der_bytes, length);
+
+  if (cert != nullptr) {
+    auto path = getKeychainPath((SecKeychainItemRef)SecCert);
+    genCertificate(cert, path, results);
+    X509_free(cert);
+  }
+
   CFRelease(der_encoded_data);
 }
 
+void genFileCertificate(const std::string& path, QueryData& results) {
+  std::string content;
+  auto s = readFile(path, content);
+  if (!s.ok()) {
+    return;
+  }
+  const unsigned char* bytes = (const unsigned char*)content.c_str();
+  X509* cert = d2i_X509(nullptr, &bytes, content.size());
+
+  if (cert != nullptr) {
+    genCertificate(cert, path, results);
+    X509_free(cert);
+  } else {
+    // If cert couldn't be read as DER, attempt
+    // to read it as a PEM.
+    BIO* bio = BIO_new_mem_buf((void*)bytes, content.size());
+
+    // There might be multiple certificates in the PEM.
+    while (PEM_read_bio_X509(bio, &cert, nullptr, nullptr) != nullptr) {
+      genCertificate(cert, path, results);
+      X509_free(cert);
+      cert = nullptr;
+    }
+
+    BIO_free(bio);
+  }
+}
+
 QueryData genCerts(QueryContext& context) {
   QueryData results;
 
-  // Allow the caller to set an explicit certificate (keychain) search path.
+  // Allow the caller to set both an explicit keychain search path
+  // and certificate files on disk.
   std::set<std::string> keychain_paths;
-  if (context.constraints["path"].exists(EQUALS)) {
-    keychain_paths = context.constraints["path"].getAll(EQUALS);
+
+  // Expand paths
+  auto paths = context.constraints["path"].getAll(EQUALS);
+  context.expandConstraints(
+      "path",
+      LIKE,
+      paths,
+      ([&](const std::string& pattern, std::set<std::string>& out) {
+        std::vector<std::string> patterns;
+        auto status =
+            resolveFilePattern(pattern, patterns, GLOB_ALL | GLOB_NO_CANON);
+        if (status.ok()) {
+          for (const auto& resolved : patterns) {
+            out.insert(resolved);
+          }
+        }
+        return status;
+      }));
+
+  if (!paths.empty()) {
+    for (const auto& path : paths) {
+      SecKeychainRef keychain = nullptr;
+      SecKeychainStatus keychain_status;
+      auto status = SecKeychainOpen(path.c_str(), &keychain);
+      if (status == 0 && keychain != nullptr &&
```

Comment:
I'm worried about this combo of conditionals. Do you mind breaking them up into subsequent checks?

Something like:
```cpp
if (status != errSecSuccess || keychain == nullptr || SecKeychainGetStatus(keychain, &keychain_status) != errSecSuccess) {
  if (keychain != nullptr) {
    CFRelease(keychain);
  }
  continue;
}
keychain_paths.insert(path);
```

works too.

We've seen crazy stuff where the return status indicates error but the output parameter still contains a result pointer for which the caller is responsible for-- combining a check for status + `nullptr` means a memory leak. Also, if `SecKeychainOpen` returns success, then `SecKeychainGetStatus` fails, there will be a memory leak.